### PR TITLE
Adding Verbose flag with new logger for better debugging

### LINKED ISSUE #51: Add debug/verbose flag to print debugging logs

### PR DESCRIPTION
**Pull request by jeffbean**
Fixes #51 by adding a --verbose or -v flag to yab to enabled Debug level logging.

Hooking up the logger also allows to use the Info level logs to show status by default.

The way I hooked this logger into the existing system was through the custom output interface made for yab. Doing this allows all the methods using output to also log more information.

### CODE REVIEW

File: benchmark.go [hunk 1]
```diff
@@ -28,6 +28,7 @@ import (
 	"sync"
 	"time"
 
+	"github.com/uber-go/zap"
```
Comment (billf) [inline benchmark.go]: this import is outside the repo, so it goes elsewhere.

File: benchmark_test.go [hunk 2]
```diff
@@ -34,6 +34,7 @@ import (
 )
 
 func TestBenchmark(t *testing.T) {
+
```
Comment (billf) [inline benchmark_test.go]: revert whitespace changes

File: integration_test.go [hunk 3]
```diff
@@ -31,6 +31,7 @@ import (
 	"testing"
 	"time"
 
+	"github.com/uber-go/zap/spy"
```
Comment (billf) [inline integration_test.go]: order

File: main_test.go [hunk 4]
```diff
@@ -31,6 +31,7 @@ import (
 	"testing"
 	"time"
 
+	"github.com/uber-go/zap/spy"
```
Comment (billf) [inline main_test.go]: import order

File: output.go [hunk 5]
```diff
@@ -25,17 +25,21 @@ import (
...
 	io.Writer
+	zap.Logger
 
 	Fatalf(format string, args ...interface{})
 	Printf(format string, args ...interface{})
 }
 
 type consoleOutput struct {
 	*os.File
+	zap.Logger
```
Comment (billf) [inline output.go]: it is now somewhat odd that `consoleOutput.Printf` and `log.Fatalf` do not use the `zap.Logger` available

Comment (jeffbean) [inline output.go]: Maybe having the Logger be a named field? that would be more explicit  
```
type consoleOutput struct {
	*os.File
	logger zap.Logger
}
```

**Comment by kriskowal** [comment]
@jeffbean I’m preparing to cut a release. Can you work with @billf to get this landed today or tomorrow?

**Comment by jeffbean** [comment]
@kriskowal I can not hit that timeline, go and cut a release if you haven't already. Thanks

File: benchmark.go [hunk 6]
```diff
@@ -28,6 +28,9 @@ import (
 	"sync"
 	"time"
 
+	"go.uber.org/zap"
```
Comment (billf) [inline benchmark.go]: these go below the yab imports

File: benchmark_test.go [hunk 7]
```diff
@@ -26,6 +26,8 @@ import (
 	"testing"
 	"time"
 
+	"go.uber.org/zap/zapcore"
```
Comment (billf) [inline benchmark_test.go]: below with testify

File: glide.yaml [hunk 8]
```diff
@@ -11,6 +11,8 @@ import:
 - package: github.com/uber/tchannel-go
   version: ^1.3
 - package: golang.org/x/net
+- package: go.uber.org/zap
+  version: ^v1
```
Comment (billf) [inline glide.yaml]: `^1`

File: logger.go [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+	// VerbosityLevelOff no log messages displayed
+	VerbosityLevelOff VerbosityLevel = iota
+	// VerbosityLevelInfo enables printing up to Info level command log statements
+	VerbosityLevelInfo
+	// VerbosityLevelDebug enables printing up to Debug level command log statements
+	VerbosityLevelDebug
+)
+
+// GetLoggerVerbosity takes a VerbosityLevel and returns an appropriate logging level
+func GetLoggerVerbosity(lvl VerbosityLevel) zapcore.Level {
```
Comment (billf) [inline logger.go]: never used outside of the package. unexport.

File: benchmark.go [hunk 10]
```diff
@@ -75,19 +77,20 @@ func (o BenchmarkOptions) enabled() bool {
 	return o.MaxDuration != 0 || o.MaxRequests != 0
 }
 
-func runWorker(t transport.Transport, m benchmarkMethod, s *benchmarkState, run *limiter.Run) {
+func runWorker(t transport.Transport, m benchmarkMethod, s *benchmarkState, run *limiter.Run, logger *zap.Logger) {
 	for cur := run; cur.More(); {
 		latency, err := m.call(t)
 		if err != nil {
 			s.recordError(err)
+			logger.Info("error occured", zap.Error(err))
```
Comment (prashantv) [inline benchmark.go]: Let's use full sentences in log messages, e.g., `Error occurred.`

File: logger.go [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+// VerbosityLevel maps to a level of logs to be enabled on the command
+type VerbosityLevel uint8
+
+const (
+	// VerbosityLevelOff no log messages displayed
+	VerbosityLevelOff VerbosityLevel = iota
```
Comment (prashantv) [inline logger.go]: we could make this map to the equivalent zap values so that we don't need any conversion functions

File: logger.go [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+		zlvl = zap.WarnLevel
+	case VerbosityLevelInfo:
+		zlvl = zap.InfoLevel
+	}
+	return zlvl
+}
+
+func configureLoggerConfig(opts *Options) zap.Config {
+	loggerConfig := zap.NewDevelopmentConfig()
+	loggerConfig.EncoderConfig = zapcore.EncoderConfig{
```
Comment (prashantv) [inline logger.go]: this matches what the default development config is already, not sure we need to set them again

File: utils_for_test.go [hunk 13]
```diff
@@ -84,3 +85,8 @@ func writeFile(t *testing.T, prefix, contents string) string {
 func getTestTracer(serviceName string) (opentracing.Tracer, io.Closer) {
 	return jaeger.NewTracer(serviceName, jaeger.NewConstSampler(true), jaeger.NewNullReporter())
 }
+
+func getTestLogger() *zap.Logger {
+	logger := zap.NewNop()
+	return logger
```
Comment (prashantv) [inline utils_for_test.go]: can inline, in fact it might not be worth having this function unless we precreate one and use it everywhere.

**Comment by prashantv** [comment]
Addressed my comments, when you get a chance @jeffbean take a look

**Comment by jeffbean** [comment]
Github won't let me approve my own RP. :D 

Changes look great. Thanks for helping and being patient with me @prashantv

**Pull request merged**